feat: add Sidebar

r1ck12 opened the pull request:
# Overview

This PR adds two components: Sidebar and LinksGroup.

* `Sidebar` is just a wrapper component using the Navbar from Mantine with basic styling.
* `LinksGroup` is a component that accepts nested links to render them as a nested collapsable menu. It can be used anywhere in the code

# Type of change

- [ ] Bug fix (non-breaking change; fixes an issue).
- [x] New feature (non-breaking change; adds functionality).
- [ ] Breaking change (fix or feature that breaks existing functionality).
- [ ] This change requires a documentation update.
- [ ] Operational change.
- [ ] Test Coverage.

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings


# Evidence
![Screen Shot 2023-05-09 at 17 00 07](https://github.com/pokt-foundation/pocket-portal/assets/90636779/6b7a61aa-45d7-4db4-8a14-52df6afb575e)

RabeeAbuBaker commented on app/components/LinksGroup/LinksGroup.tsx:
```diff
@@ -0,0 +1,110 @@
+          }}
+        >
+          <Box sx={{ display: "flex", alignItems: "center" }}>
+            <Text
```
This component and the items component above have a lot in common. Can we group them into a single component to reduce repetition and allow it to be customized (still accept outer props when needed)

RabeeAbuBaker commented:
It looks great 👏
But one thing I just wanted to double check is the possibility of having more nested links where children also have children. Is that a possibility?

r1ck12 commented:
> It looks great 👏 But one thing I just wanted to double check is the possibility of having more nested links where children also have children. Is that a possibility?

According to the types definition yeah, but I'll make sure the UI can handle that too

PseudoMera reviewed (approved):
Other than @RabeeAbuBaker comments, everything looks good!

r1ck12 commented:
@RabeeAbuBaker @PseudoMera I made the `LinksGroup` recursive to render nested children links and added the `LinkItem` component to reduce code repetition

RabeeAbuBaker commented on app/components/LinksGroup/LinksGroup.tsx:
```diff
@@ -11,18 +11,55 @@ import {
+}: LinkItemProps) => (
+  <Text
+    m={0}
+    p={nesting_level ? `10.5px 0 10.5px ${nesting_level * 32}px` : "16px 8px"}
```
Nice!

RabeeAbuBaker reviewed (approved):
Great work! Loved the recursive implementation 👏

r1ck12 commented:
@RabeeAbuBaker @PseudoMera I added the Sidebar to the docs page, please review it again when you have time 🫶

RabeeAbuBaker commented on app/routes/_landing.($lang).docs/route.tsx:
```diff
@@ -27,18 +31,44 @@ export const loader: LoaderFunction = async ({ params }) => {
-    </div>
+    <Grid
+      gutter="md"
+      sx={{
```
Just FYI... I have added Flex to blocks. Let's try to use it when possible in the future.

RabeeAbuBaker commented on app/root.tsx:
```diff
@@ -280,7 +282,14 @@ export default function App() {
+          <Container
+            className="container"
+            size="lg"
+            sx={{
```
I think it's better to use `ml` and `mr` props directly?

RabeeAbuBaker commented on app/routes/_landing.($lang).docs/route.tsx:
```diff
@@ -13,7 +17,7 @@ export const loader: LoaderFunction = async ({ params }) => {
   try {
     const doc = await cms.getDocs({
-      filter: { status: { _eq: "published" } },
+      // filter: { status: { _eq: "published" } },
```
leftover?

RabeeAbuBaker reviewed (commented):
Do we still need `replaceAll`?

RabeeAbuBaker commented on app/routes/_landing.($lang).docs/route.tsx:
```diff
@@ -30,31 +30,63 @@ export const loader: LoaderFunction = async ({ params }) => {
+    docs.forEach((doc) => {
+      if (doc.parent) {
+        let parentDoc = doc.parent
+        const parentInMap = mappedDocs.find((item) => item.id === parentDoc.id)
```
for this case I thing `some()` is better?

RabeeAbuBaker commented on app/routes/_landing.($lang).docs/route.tsx:
```diff
@@ -30,31 +30,63 @@ export const loader: LoaderFunction = async ({ params }) => {
+  const organizeData = (docs: documentation[]): LinksGroupProps[] => {
+    let mappedDocs: LinksGroupProps[] = []
+
+    docs.forEach((doc) => {
```
imo it's better to use `map` here and avoid pushing. Also, for the usage of the variable below, instead of replacing the array with a new version of itself using map we can create a new const and return it. 
The code this way stays more readable and easy to understand.

RabeeAbuBaker commented on app/routes/_landing.($lang).docs/route.tsx:
```diff
@@ -30,31 +30,63 @@ export const loader: LoaderFunction = async ({ params }) => {
+      const organizedData = organizeData(data)
+      setLinksGroupItems(organizedData)
     }
   }, [data])
```
`organizeData` is missing from the dependency array. We can memoize the function and pass it here, or a more common approach it can be assigned to a ref and used via the ref which wouldn't be require it to be added to the dependency array.

RabeeAbuBaker commented:
Just a quick thought... Do we need to keep track of nesting levels and all of the data transformation done above? 
Can we instead use https://mantine.dev/core/list (removing the bullets with css) and nest levels where the children will automatically have spacing?

Alternatively, we could just wrap children [items](https://github.com/pokt-foundation/pocket-portal/pull/275/files#diff-05faffb5603a4e3bbb6ae112127cb02fd5eed1324ab15342d43a71d8a5766947R101) in a `div` and add padding or style the font for items wrapped in it.

I'm trying to think of ways to avoid touching the data and manually setting the nesting level.

r1ck12 commented:
> Just a quick thought... Do we need to keep track of nesting levels and all of the data transformation done above? Can we instead use https://mantine.dev/core/list (removing the bullets with css) and nest levels where the children will automatically have spacing?
> 
> Alternatively, we could just wrap children [items](https://github.com/pokt-foundation/pocket-portal/pull/275/files#diff-05faffb5603a4e3bbb6ae112127cb02fd5eed1324ab15342d43a71d8a5766947R101) in a `div` and add padding or style the font for items wrapped in it.
> 
> I'm trying to think of ways to avoid touching the data and manually setting the nesting level.

am I misunderstanding what you are saying or would we also need to format the data into a nested array in any cases? Unless we receive the data as a nested array directly from the API

RabeeAbuBaker commented:
> 



> > Just a quick thought... Do we need to keep track of nesting levels and all of the data transformation done above? Can we instead use https://mantine.dev/core/list (removing the bullets with css) and nest levels where the children will automatically have spacing?
> > Alternatively, we could just wrap children [items](https://github.com/pokt-foundation/pocket-portal/pull/275/files#diff-05faffb5603a4e3bbb6ae112127cb02fd5eed1324ab15342d43a71d8a5766947R101) in a `div` and add padding or style the font for items wrapped in it.
> > I'm trying to think of ways to avoid touching the data and manually setting the nesting level.
> 
> am I misunderstanding what you are saying or would we also need to format the data into a nested array in any cases? Unless we receive the data as a nested array directly from the API

Oh I was not aware of that. In this case, either way we'll have to format the data, but we can rethink of the nesting level usage an setting styling based on its value, I think it's not needed since we can add padding to items as I mentioned above.

r1ck12 commented:
> > 
> 
> > > Just a quick thought... Do we need to keep track of nesting levels and all of the data transformation done above? Can we instead use https://mantine.dev/core/list (removing the bullets with css) and nest levels where the children will automatically have spacing?
> > > Alternatively, we could just wrap children [items](https://github.com/pokt-foundation/pocket-portal/pull/275/files#diff-05faffb5603a4e3bbb6ae112127cb02fd5eed1324ab15342d43a71d8a5766947R101) in a `div` and add padding or style the font for items wrapped in it.
> > > I'm trying to think of ways to avoid touching the data and manually setting the nesting level.
> > 
> > 
> > am I misunderstanding what you are saying or would we also need to format the data into a nested array in any cases? Unless we receive the data as a nested array directly from the API
> 
> Oh I was not aware of that. In this case, either way we'll have to format the data, but we can rethink of the nesting level usage an setting styling based on its value, I think it's not needed since we can add padding to items as I mentioned above.

I tried both of your solutions and here's what I got:

1. Using the Mantine `List` and `List.Item` I could not set the `withPadding` prop (it doesn't exist in the types definition) I believe this because we don't have an upgraded version of Mantine on our pocket-blocks repo. 

2. I tried wrapping the items around and setting a `padding` attribute but we still have to increment the level to set the padding according to the item nesting level.

In the end, I removed the `nested_level` logic from the router, and replaced the `Grid` with the `List`. I would like to upgrade our Mantine later and revisit this to use the `List.Item` `withPadding` so we won't need the level logic

RabeeAbuBaker commented:
> > > 
> > 
> > 
> > > > Just a quick thought... Do we need to keep track of nesting levels and all of the data transformation done above? Can we instead use https://mantine.dev/core/list (removing the bullets with css) and nest levels where the children will automatically have spacing?
> > > > Alternatively, we could just wrap children [items](https://github.com/pokt-foundation/pocket-portal/pull/275/files#diff-05faffb5603a4e3bbb6ae112127cb02fd5eed1324ab15342d43a71d8a5766947R101) in a `div` and add padding or style the font for items wrapped in it.
> > > > I'm trying to think of ways to avoid touching the data and manually setting the nesting level.
> > > 
> > > 
> > > am I misunderstanding what you are saying or would we also need to format the data into a nested array in any cases? Unless we receive the data as a nested array directly from the API
> > 
> > 
> > Oh I was not aware of that. In this case, either way we'll have to format the data, but we can rethink of the nesting level usage an setting styling based on its value, I think it's not needed since we can add padding to items as I mentioned above.
> 
> I tried both of your solutions and here's what I got:
> 
> 1. Using the Mantine `List` and `List.Item` I could not set the `withPadding` prop (it doesn't exist in the types definition) I believe this because we don't have an upgraded version of Mantine on our pocket-blocks repo.
> 2. I tried wrapping the items around and setting a `padding` attribute but we still have to increment the level to set the padding according to the item nesting level.
> 
> In the end, I removed the `nested_level` logic from the router, and replaced the `Grid` with the `List`. I would like to upgrade our Mantine later and revisit this to use the `List.Item` `withPadding` so we won't need the level logic

Got it...

I gave it a go and tried to wrap the children in a Box and give it the same padding (because padding will accumulate -child padding + parent padding-) and it worked without needing the nesting level. 

![image](https://github.com/pokt-foundation/pocket-portal/assets/38038348/1453b54e-232b-43bf-b95c-03299a70125f)

The change I made:

![image](https://github.com/pokt-foundation/pocket-portal/assets/38038348/b480014d-116b-4b92-bd44-b90a79510f49)

RabeeAbuBaker commented on app/components/LinksGroup/LinksGroup.tsx:
```diff
@@ -0,0 +1,152 @@
+      textTransform: "capitalize",
+    }}
+  >
+    <a href={link}>{label}</a>
```
I think we can improve the performance if we use `Link` component from Remix, and use the prefetch prop with "intent". Navigation will be much faster and feel snappier. 

https://remix.run/docs/en/main/components/link#prefetch

r1ck12 commented:
@RabeeAbuBaker That makes sense thanks for the code explanation, I'll get back into this in a few mins
